File: tarot/payments.py
"""Payment providers, behind one interface.

The single most important decision in this file is that there *is* a file. This
project sells something in a category that payment processors restrict, so the
provider will change at least once — under duress, probably, with live orders in
the queue. Switching has to be a config change, not a refactor.

Two providers ship:

  * ``manual``  — settlement happens out of band. No network calls, no keys, and
    the whole flow works end to end. This is the default and it is not a stub.
  * ``waffo``   — Waffo as Merchant of Record.

**The Waffo adapter is written against an unverified specification.** The
documentation hosts (docs.waffo.com, docs.waffo.ai) are unreachable from the
build environment, so the request shape below comes from public secondary
sources, not from reading the API reference. Every point that needs confirming
is marked ``UNVERIFIED`` and listed in ``WAFFO_OPEN_QUESTIONS``. Run

    python payments.py check

against a real key before trusting any of it. Until that passes, leave
``PAYMENT_PROVIDER=manual``.

Two questions were put to Waffo and answered:

  * **Category.** Tarot and spiritual content are permitted — with WeChat Pay
    excluded. That exclusion is enforced here (``WAFFO_EXCLUDED_METHODS``) rather
    than left to the dashboard, so it travels with the code and is visible in
    review. It also means mainland Chinese buyers lose their dominant method,
    which is an argument for leading with the English-language market.
  * **Splits.** Waffo does not support paying a third party a share. It settles
    the full amount to one payee, so the site is the seller and readers are
    subcontractors invoicing it — a studio, not a marketplace. What each reader
    is owed is tracked in ``store.py`` and paid out of band.
"""

from __future__ import annotations

import hashlib
import hmac
import json
import os
import sys
import urllib.error
import urllib.request
from dataclasses import dataclass

# Statuses an order's payment can be in. Anything else is a bug.
UNPAID = "unpaid"
PENDING = "pending"
PAID = "paid"
REFUNDED = "refunded"
FAILED = "failed"

PAYMENT_STATUSES = {UNPAID, PENDING, PAID, REFUNDED, FAILED}


@dataclass
class Checkout:
    """Where to send the buyer, and what to remember about it."""
    url: str
    reference: str
    status: str = PENDING


class PaymentError(RuntimeError):
    pass


# ---------------------------------------------------------------------------
# Manual settlement
# ---------------------------------------------------------------------------

class ManualProvider:
    """Records the intent to pay; money moves elsewhere.

    This is what runs while the provider question is open, and it is a complete
    path rather than a placeholder: an order is created, an operator confirms
    payment, and the reading proceeds.
    """

    name = "manual"
    is_live = False

    def __init__(self, config):
        self.config = config

    def create_checkout(self, order, return_url, cancel_url):
        return Checkout(url=return_url, reference=f"manual:{order['token']}",
                        status=UNPAID)

    def verify_webhook(self, headers, raw_body):
        # Nothing sends webhooks here. Refusing is the correct answer, not an
        # oversight: an endpoint that accepts unsigned "paid" events is a way to
        # get free readings.
        return None


# ---------------------------------------------------------------------------
# Waffo (Merchant of Record)
# ---------------------------------------------------------------------------

WAFFO_OPEN_QUESTIONS = [
    "Base URL: api.waffo.ai vs api.waffo.com — waffo.com and waffo.ai both "
    "appear to be Waffo properties and may be different products.",
    "Auth: whether the API key goes in Authorization: Bearer, or a bespoke "
    "header, and whether requests must additionally be signed.",
    "Whether X-Store-Slug and X-Environment are required on every call.",
    "The create-session response field holding the redirect URL.",
    "Webhook signature: header name, algorithm, and whether the signed payload "
    "is the raw body alone or is prefixed with a timestamp.",
    "Webhook event names for a completed payment and for a refund.",
    "The field name for excluding a payment method — excludedPaymentMethods "
    "is a guess, and getting it wrong silently re-enables WeChat Pay.",
]


class WaffoProvider:
    """Waffo as Merchant of Record.

    UNVERIFIED throughout — see the module docstring and WAFFO_OPEN_QUESTIONS.
    Correct the four constants below once the reference is in hand; nothing
    outside this class needs to change.
    """

    name = "waffo"
    is_live = True

    # --- the four things to correct against the real reference ---------------
    BASE_URL = "https://api.waffo.ai/v1"                      # UNVERIFIED
    CREATE_SESSION = "/actions/checkout/create-session"        # UNVERIFIED
    SIGNATURE_HEADER = "X-Waffo-Signature"                     # UNVERIFIED
    PAID_EVENTS = {"checkout.completed", "payment.succeeded"}  # UNVERIFIED
    REFUND_EVENTS = {"payment.refunded", "charge.refunded"}    # UNVERIFIED
    # -------------------------------------------------------------------------

    def __init__(self, config):
        self.api_key = config.get("WAFFO_API_KEY", "")
        self.store_slug = config.get("WAFFO_STORE_SLUG", "")
        self.environment = config.get("WAFFO_ENVIRONMENT", "sandbox")
        self.webhook_secret = config.get("WAFFO_WEBHOOK_SECRET", "")
        self.product_id = config.get("WAFFO_PRODUCT_ID", "")
        # Waffo permit this category on the condition that WeChat Pay is not
        # offered. Enforced in the request, not only in their dashboard.
        self.excluded_methods = [
            m.strip() for m in
            (config.get("WAFFO_EXCLUDED_METHODS") or "wechat_pay").split(",")
            if m.strip()
        ]
        if not self.api_key:
            raise PaymentError(
                "PAYMENT_PROVIDER=waffo but WAFFO_API_KEY is unset. "
                "Set it, or fall back to PAYMENT_PROVIDER=manual."
            )

    def _headers(self):
        return {
            "Authorization": f"Bearer {self.api_key}",   # UNVERIFIED
            "Content-Type": "application/json",
            "X-Store-Slug": self.store_slug,             # UNVERIFIED
            "X-Environment": self.environment,           # UNVERIFIED
        }

    def _post(self, path, payload, timeout=20):
        req = urllib.request.Request(
            self.BASE_URL + path,
            data=json.dumps(payload).encode(),
            headers=self._headers(),
            method="POST",
        )
        try:
            with urllib.request.urlopen(req, timeout=timeout) as resp:
                return json.loads(resp.read().decode() or "{}")
        except urllib.error.HTTPError as exc:
            body = exc.read().decode(errors="replace")[:500]
            raise PaymentError(f"Waffo {path} -> HTTP {exc.code}: {body}") from exc
        except (urllib.error.URLError, TimeoutError, OSError) as exc:
            raise PaymentError(f"Waffo {path} unreachable: {exc}") from exc

    def create_checkout(self, order, return_url, cancel_url):
        payload = {                                      # UNVERIFIED field names
            "productId": self.product_id,
            "productType": "one_time",
            "currency": order["currency"],
            "amount": order["price_cents"],
            "referenceId": order["token"],
            "successUrl": return_url,
            "cancelUrl": cancel_url,
            "excludedPaymentMethods": self.excluded_methods,   # UNVERIFIED name
        }
        data = self._post(self.CREATE_SESSION, payload)

        url = _first(data, "checkoutUrl", "url", "redirectUrl", "paymentUrl")
        ref = _first(data, "sessionId", "id", "sessionToken") or order["token"]
        if not url:
            raise PaymentError(
                f"Waffo create-session returned no redirect URL. Keys present: "
                f"{sorted(data)[:12]}. Correct the field name in WaffoProvider."
            )
        return Checkout(url=url, reference=str(ref), status=PENDING)

    def verify_webhook(self, headers, raw_body):
        """Return (reference, payment_status) or None.

        Fails closed: without a configured secret, or on any signature mismatch,
        the event is discarded. A webhook that can be forged is a way to get
        paid-tier work for free.
        """
        if not self.webhook_secret:
            return None
        sent = headers.get(self.SIGNATURE_HEADER, "")
        if not sent:
            return None

        expected = hmac.new(self.webhook_secret.encode(), raw_body,
                            hashlib.sha256).hexdigest()          # UNVERIFIED
        if not hmac.compare_digest(sent.strip().lower(), expected):
            return None

        try:
            event = json.loads(raw_body.decode())
        except (ValueError, UnicodeDecodeError):
            return None

        kind = str(_first(event, "type", "event", "eventType") or "")
        data = event.get("data") or event
        reference = _first(data, "referenceId", "reference", "metadata_reference")
        if not reference:
            return None

        if kind in self.PAID_EVENTS:
            return str(reference), PAID
        if kind in self.REFUND_EVENTS:
            return str(reference), REFUNDED
        return None


def _first(mapping, *keys):
    """First present, non-empty value among `keys`. Tolerates naming drift."""
    if not isinstance(mapping, dict):
        return None
    for k in keys:
        v = mapping.get(k)
        if v:
            return v
    return None


# ---------------------------------------------------------------------------
# Registry
# ---------------------------------------------------------------------------

PROVIDERS = {"manual": ManualProvider, "waffo": WaffoProvider}


def get_provider(config):
    name = (config.get("PAYMENT_PROVIDER") or "manual").strip().lower()
    if name not in PROVIDERS:
        raise PaymentError(
            f"Unknown PAYMENT_PROVIDER {name!r}. Known: {', '.join(sorted(PROVIDERS))}"
        )
    return PROVIDERS[name](config)


# ---------------------------------------------------------------------------
# Live check
#
# The point of this command is to find out which of WAFFO_OPEN_QUESTIONS were
# guessed wrong, using one real request, before any customer sees a checkout.
# ---------------------------------------------------------------------------

def check(config=None):
    config = config or dict(os.environ)
    print("Waffo adapter check\n" + "=" * 60)
    print("The request shape below is UNVERIFIED. Open questions:")
    for i, q in enumerate(WAFFO_OPEN_QUESTIONS, 1):
        print(f"  {i}. {q}")
    print("=" * 60)

    missing = [k for k in ("WAFFO_API_KEY", "WAFFO_STORE_SLUG", "WAFFO_PRODUCT_ID")
               if not config.get(k)]
    if missing:
        print(f"\nNot configured: {', '.join(missing)}")
        print("Set them and re-run to make a real request.")
        return 1

    provider = WaffoProvider(config)
    print(f"\nPOST {provider.BASE_URL}{provider.CREATE_SESSION}")
    print(f"  environment: {provider.environment}")
    print(f"  excluding:   {', '.join(provider.excluded_methods) or '(nothing)'}")
    fake = {"token": "check-" + "0" * 8, "price_cents": 100, "currency": "EUR"}
    try:
        result = provider.create_checkout(
            fake, "https://example.invalid/ok", "https://example.invalid/cancel")
    except PaymentError as exc:
        print(f"\nFAILED: {exc}")
        print("\nThat error is the useful output — it names which assumption is "
              "wrong. Correct the constants in WaffoProvider and re-run.")
        return 1

    print(f"\nOK  redirect: {result.url[:80]}\n    reference: {result.reference}")
    print("\nStill unconfirmed by this check:")
    print("  - webhook signature scheme and event names — send a sandbox payment")
    print("    and inspect a real delivery.")
    print("  - that the exclusion was honoured. OPEN THE CHECKOUT PAGE and confirm")
    print("    WeChat Pay is absent. A silently ignored field looks like success")
    print("    here and breaks the category approval in production.")
    return 0


if __name__ == "__main__":
    sys.exit(check() if len(sys.argv) > 1 and sys.argv[1] == "check" else check())

File: tarot/web/apps/site/lib/payments.ts
import "server-only"

import { WaffoPancake, verifyWebhook, TaxCategory, type PaymentMethod } from "@waffo/pancake-ts"

import type { Order } from "./orders"

/** Taking money, behind one interface.
 *
 * The default provider settles by hand. That is not a stub standing in for the
 * real thing -- it is how a small studio actually starts, and it means the
 * whole order flow is walkable with no provider account at all.
 */

export interface CheckoutSession {
  url: string
  reference: string
}

export interface PaymentProvider {
  readonly name: string
  /** True when the provider can actually be called -- credentials present. */
  configured(): boolean
  createCheckout(order: Order, opts: { successUrl: string; cancelUrl: string }):
    Promise<CheckoutSession>
  /** Verify a webhook and say what it means. Returns null if not verified. */
  verifyWebhook(body: string, headers: Headers):
    { event: string; reference: string } | null
}

/** The default, and a complete path rather than a stub.
 *
 * The operator marks orders paid in /admin. That is genuinely how a small
 * studio starts -- bank transfer, invoice, whatever -- and it means the whole
 * flow is walkable with no provider account at all.
 */
class ManualProvider implements PaymentProvider {
  readonly name = "manual"
  configured() {
    return true
  }
  async createCheckout(order: Order): Promise<CheckoutSession> {
    return { url: `/order/${order.token}?awaiting=1`, reference: order.token }
  }
  verifyWebhook() {
    return null
  }
}

// ---------------------------------------------------------------------------
// Waffo Pancake
//
// Waffo is a merchant of record: it sells to the customer, handles the tax, and
// settles to us. That is the point of it for a Chinese entity selling to
// customers abroad -- the alternative is registering for VAT in every country a
// customer happens to live in.
//
// Everything below is the SDK's documented shape rather than a guess. Four
// things about it are not what a general payments intuition would predict, and
// each one is a silent failure if you get it wrong:
//
//  1. Auth is RSA-SHA256 request signing with a merchant id and a PEM private
//     key -- not a bearer API key. The key never leaves the server.
//  2. Checkout charges a *product*, created in Waffo's dashboard. A reader sets
//     their own price, so one product per price would mean a new product every
//     time someone edits their rate, and a missing one would charge whatever
//     the product happens to say. `priceSnapshot` overrides the amount per
//     session instead, which is why there is exactly one product id below.
//  3. The webhook signature is asymmetric: Waffo signs with its private key and
//     we verify with its public one. There is no shared secret to configure and
//     nothing for us to keep -- an HMAC scheme here would be inventing a secret
//     that does not exist.
//  4. `orderMerchantExternalId` is inherited by every payment and refund on the
//     order, which is what lets a refund three weeks later still name the order
//     it belongs to.
// ---------------------------------------------------------------------------

/** Payment methods that must not be offered.
 *
 * Waffo's answer was that the category is allowed but WeChat Pay is not
 * available for it. Encoded here rather than in a runbook, because a runbook
 * does not stop the next deploy from re-enabling it.
 *
 * On EUR this is belt-and-braces: WeChat is a CNY-only method in Waffo's
 * matrix, so the currency already excludes it. It matters the day someone adds
 * a CNY price.
 */
export const WAFFO_EXCLUDED_METHODS = (
  process.env.WAFFO_EXCLUDED_METHODS ?? "wechat"
)
  .split(",")
  .map((s) => s.trim())
  .filter(Boolean) as PaymentMethod[]

/** A tarot reading is a service someone writes, not a downloaded file.
 *
 * This drives the tax Waffo charges the customer, so it is a business answer
 * rather than a technical one -- overridable, but not silently.
 */
const WAFFO_TAX_CATEGORY = (process.env.WAFFO_TAX_CATEGORY ??
  TaxCategory.ProfessionalService) as TaxCategory

const WAFFO_SIGNATURE_HEADER = "x-waffo-signature"

/** Cents to the display string Waffo prices in ("35.00", not 3500).
 *
 * Its own name for the field is `amount: string`, and handing it 3500 would be
 * accepted -- as thirty-five hundred euros. Nothing errors, the customer sees
 * the wrong number at the last possible moment, and the only signal is a
 * chargeback. Worth a named function so it can be checked.
 */
export function displayAmount(cents: number): string {
  return (cents / 100).toFixed(2)
}

/** Events that mean money moved, and which way.
 *
 * `subscription.*` are here because the same product can later be sold as a
 * standing monthly reading; on a one-off order they simply never arrive.
 */
const WAFFO_PAID_EVENTS = new Set([
  "order.completed",
  "subscription.activated",
  "subscription.payment_succeeded",
])
const WAFFO_REFUND_EVENTS = new Set(["refund.succeeded"])

class WaffoProvider implements PaymentProvider {
  readonly name = "waffo"

  configured() {
    return Boolean(
      process.env.WAFFO_MERCHANT_ID &&
        process.env.WAFFO_PRIVATE_KEY &&
        process.env.WAFFO_PRODUCT_ID,
    )
  }

  private client(): WaffoPancake {
    return new WaffoPancake({
      merchantId: process.env.WAFFO_MERCHANT_ID!,
      privateKey: process.env.WAFFO_PRIVATE_KEY!,
      ...(process.env.WAFFO_BASE_URL ? { baseUrl: process.env.WAFFO_BASE_URL } : {}),
      ...(process.env.WAFFO_WEBHOOK_PUBLIC_KEY
        ? { webhookPublicKey: process.env.WAFFO_WEBHOOK_PUBLIC_KEY }
        : {}),
    })
  }

  async createCheckout(
    order: Order,
    opts: { successUrl: string; cancelUrl: string },
  ): Promise<CheckoutSession> {
    if (!this.configured()) {
      throw new Error(
        "Waffo is not configured (needs WAFFO_MERCHANT_ID, WAFFO_PRIVATE_KEY, WAFFO_PRODUCT_ID)",
      )
    }

    const session = await this.client().checkout.anonymous.create({
      productId: process.env.WAFFO_PRODUCT_ID!,
      currency: order.currency,
      // The amount the customer actually sees on the reader's page. Without
      // this the product's own price is charged, and the mismatch shows up as
      // a chargeback rather than an error.
      priceSnapshot: {
        amount: displayAmount(order.price_cents),
        taxCategory: WAFFO_TAX_CATEGORY,
      },
      successUrl: opts.successUrl,
      // Inherited by every payment and refund on this order, so a refund weeks
      // later still names the order it belongs to.
      orderMerchantExternalId: order.token,
      metadata: { token: order.token, reader: order.reader_slug },
      ...(WAFFO_EXCLUDED_METHODS.length
        ? { excludePaymentMethods: WAFFO_EXCLUDED_METHODS }
        : {}),
    })

    // The session id, not the order token: the token is already the order's
    // primary key, and storing it as the payment reference would tell us
    // nothing new. orderByPaymentRef matches either.
    return { url: session.checkoutUrl, reference: session.sessionId }
  }

  verifyWebhook(body: string, headers: Headers) {
    // Asymmetric: Waffo signs `${t}.${rawBody}` with its private key and we
    // check it against the public one the SDK ships. Throws on a bad
    // signature, a missing header, or a timestamp outside the tolerance --
    // all of which mean the same thing here.
    let event
    try {
      event = verifyWebhook(body, headers.get(WAFFO_SIGNATURE_HEADER))
    } catch {
      return null
    }

    const data = (event.data ?? {}) as {
      orderMerchantExternalId?: string
      orderMetadata?: Record<string, string>
    }
    // Order-level and inherited, so this is set on refunds too. The metadata
    // copy is the fallback for the same reason it exists in the first place.
    const reference = data.orderMerchantExternalId || data.orderMetadata?.token || ""
    if (!reference) return null

    if (WAFFO_PAID_EVENTS.has(event.eventType)) return { event: "paid", reference }
    if (WAFFO_REFUND_EVENTS.has(event.eventType)) return { event: "refunded", reference }
    // Everything else -- past_due, plan changes, cancellations -- is a real
    // event we simply have no state for. Not an error.
    return null
  }
}

const PROVIDERS: Record<string, PaymentProvider> = {
  manual: new ManualProvider(),
  waffo: new WaffoProvider(),
}

export function provider(): PaymentProvider {
  const chosen = process.env.PAYMENT_PROVIDER ?? "manual"
  return PROVIDERS[chosen] ?? PROVIDERS.manual
}

/** What /admin shows. A provider selected but not configured is the failure
 *  worth surfacing: checkout throws, and it throws at the worst moment. */
export function paymentStatus() {
  const chosen = process.env.PAYMENT_PROVIDER ?? "manual"
  const active = provider()
  return { requested: chosen, active: active.name, configured: active.configured() }
}
